plugins/linux-capture/pipewire.c

```diff
@@ -367,6 +377,155 @@ static void swap_texture_red_blue(gs_texture_t *texture)
+	 * contain. In this case we enumerate supported formats. */
+	spa_pod_builder_push_object(b, &f[0], SPA_TYPE_OBJECT_Format,
+				    SPA_PARAM_EnumFormat);
+	/* add media type and media subtype properties */
+	spa_pod_builder_add(b, SPA_FORMAT_mediaType,
+			    SPA_POD_Id(SPA_MEDIA_TYPE_video), 0);
+	spa_pod_builder_add(b, SPA_FORMAT_mediaSubtype,
+			    SPA_POD_Id(SPA_MEDIA_SUBTYPE_raw), 0);
+
+	/* formats */
+	spa_pod_builder_add(b, SPA_FORMAT_VIDEO_format, SPA_POD_Id(format), 0);
+
+	/* modifier */
+	if (modifier_count > 0) {
+		/* build an enumeration of modifiers */
+		spa_pod_builder_prop(b, SPA_FORMAT_VIDEO_modifier,
+				     SPA_POD_PROP_FLAG_MANDATORY |
+					     SPA_POD_PROP_FLAG_DONT_FIXATE);
+		spa_pod_builder_push_choice(b, &f[1], SPA_CHOICE_Enum, 0);
+		/* modifiers from  an array */
+		for (uint32_t i = 0; i < modifier_count; i++) {
+			uint64_t modifier = modifiers[i];
+			spa_pod_builder_long(b, modifier);
+			if (i == 0)
+				spa_pod_builder_long(b, modifier);
+		}
+		spa_pod_builder_pop(b, &f[1]);
+	}
+	/* add size and framerate ranges */
+	spa_pod_builder_add(b, SPA_FORMAT_VIDEO_size,
+			    SPA_POD_CHOICE_RANGE_Rectangle(
+				    &SPA_RECTANGLE(320, 240), // Arbitrary
+				    &SPA_RECTANGLE(1, 1),
+				    &SPA_RECTANGLE(8192, 4320)),
+			    SPA_FORMAT_VIDEO_framerate,
+			    SPA_POD_CHOICE_RANGE_Fraction(
+				    &SPA_FRACTION(ovi->fps_num, ovi->fps_den),
+				    &SPA_FRACTION(0, 1), &SPA_FRACTION(360, 1)),
+			    0);
+	return spa_pod_builder_pop(b, &f[0]);
+}
+
+static uint32_t build_format_params(obs_pipewire_data *obs_pw,
+				    struct spa_pod_builder *pod_builder,
+				    const struct spa_pod ***param_list)
+{
+	uint32_t n_params = 0;
+
+	const struct spa_pod **params;
+	params = bzalloc(2 * obs_pw->n_formats * sizeof(struct spa_pod *));
+
+	for (int i = 0; i < obs_pw->n_formats; i++) {
+		if (obs_pw->modifier_info[i].n_modifiers == 0) {
+			continue;
+		}
+		params[n_params++] =
+			build_format(pod_builder, &obs_pw->video_info,
+				     obs_pw->modifier_info[i].spa_format,
+				     obs_pw->modifier_info[i].modifiers,
+				     obs_pw->modifier_info[i].n_modifiers);
+	}
+	for (int i = 0; i < obs_pw->n_formats; i++) {
+		params[n_params++] = build_format(
+			pod_builder, &obs_pw->video_info,
+			obs_pw->modifier_info[i].spa_format, NULL, 0);
+	}
+	*param_list = params;
+	return n_params;
+}
+
+static uint32_t create_modifier_info(struct modifier_info **modifier_info)
+{
+	uint32_t formats[] = {
+		SPA_VIDEO_FORMAT_BGRA,
+		SPA_VIDEO_FORMAT_RGBA,
+		SPA_VIDEO_FORMAT_BGRx,
+		SPA_VIDEO_FORMAT_RGBx,
+	};
+
+	size_t n_formats = sizeof(formats) / sizeof(formats[0]);
+
+	struct modifier_info *info =
+		bzalloc(n_formats * sizeof(struct modifier_info));
+
+	obs_enter_graphics();
+
+	enum gs_dmabuf_flags dmabuf_flags;
+	uint32_t *drm_formats = NULL;
+	size_t n_drm_formats;
+	bool query_capabilities = gs_query_dmabuf_capabilities(
+		&dmabuf_flags, &drm_formats, &n_drm_formats);
```

Comment:
Probably easier to follow with an explicit success / failure separation:

```c
if (!gs_query_dmabuf_capabilities(&dmabuf_flags, &drm_formats, &n_drm_formats)) {
    // Failed to query capabilities, assume the following formats work
   ...
   return false;
}
```

Reply:
Not really we still have to iterate over all formats, since we generate both DMA-BUF and SHM pipewire params from this list.

Reply:
Ok, to make it more clear: I renamed it from modifier_info to format_info, since this stores all information about a given format OBS can use and is used to generate all FormatParams for PipeWire. So we have to create and fill a struct for each format. We only can skip the modifier part if nothing is available.

Reply:
Okay, so could you please rename the variable to `capabilities_queried`, in past tense, so the if condition is slightly easier to read?